platforms/core-configuration/kotlin-dsl-provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/KotlinDslDclSchemaCollector.kt

```diff
@@ -16,36 +16,136 @@
+            val analysisSchema = step.evaluationSchemaForStep.analysisSchema
+            val typeRefContext = SchemaTypeRefContext(analysisSchema)
+            val types = analysisSchema.dataClassTypesByFqName.values.filterIsInstance<DataClass>()
+            types.flatMap { type ->
+                type.memberFunctions.mapNotNull { function ->
+                    function.metadata.dataOfTypeOrNull<ContainerElementFactory>()?.let { factory ->
+                        val containerType = typeOf(type, classLoader)
+                            ?: return@let null
+                        val elementDclType = typeRefContext.resolveRef(factory.elementType) as? DataClass
+                            ?: return@let null
+                        val elementType = typeOf(elementDclType, classLoader)
+                            ?: return@let null
+                        ContainerElementFactoryEntry<TypeOf<*>>(function.simpleName, containerType, elementType)
+                    }
+                }
+            }
+        }
+    }
 
-        val elementTypes = getters.mapNotNullTo(hashSetOf()) {
-            DclContainerMemberExtractionUtils.elementTypeFromNdocContainerType(it.genericReturnType)
+    private fun typeOf(dclDataClass: DataClass, classLoader: ClassLoader): TypeOf<*>? {
+        fun loadClassOrNull(className: String): Class<*>? = try {
+            classLoader.loadClass(className)
+        } catch (_: ClassNotFoundException) {
+            null
         }
 
-        return elementTypes.map { type ->
-            val elementType = TypeOf.typeOf<Any>(type)
-            val scopeReceiverType = TypeOf.parameterizedTypeOf(typeOf<NamedDomainObjectContainer<*>>(), elementType)
-            val factoryName = DclContainerMemberExtractionUtils.elementFactoryFunctionNameFromElementType(elementType.concreteClass.kotlin)
-            ContainerElementFactoryEntry(factoryName, scopeReceiverType, elementType)
+        val loadedClass = loadClassOrNull(dclDataClass.javaTypeName)
+            ?: return null
+
+        val typeArgs = dclDataClass.javaTypeArgumentTypeNames.map {
+            loadClassOrNull(it) ?: return null
+        }
+
+        return if (typeArgs.isNotEmpty()) {
+            /**
+             * Workaround: The [TypeOf] infrastructure handles parameterized types specially.
+             * Passing the raw [Class] obtained from the class loader to [TypeOf.parameterizedTypeOf] would not work.
+             * We need to provide a [ParameterizedType] instance.
+             */
+            TypeOf.typeOf<Any>(object : ParameterizedType {
+                override fun getActualTypeArguments(): Array<Type> = typeArgs.toTypedArray<Type>()
+                override fun getRawType(): Type = loadedClass
+
+                /** [Class.getNestHost] is @since 11, cannot use it; but we are fine with no owner type here. */
+                override fun getOwnerType() = null
+            })
+        } else{
+            TypeOf.typeOf(loadedClass)
         }
```

Comment:
💭  We could consider extracting this as a utility factory function in `TypeOf` for such cases

Reply:
Done in [`d8d817d` (#31042)](https://github.com/gradle/gradle/pull/31042/commits/d8d817d2d726d844d4b4e667455914bcac2ae57e)